relay-server/src/utils/dynamic_sampling.rs

```diff
@@ -1,152 +1,228 @@
-    /// A set of Glob patterns for which the rule applies, empty applies to all
-    #[serde(default)]
-    pub releases: GlobPatterns,
-    /// A set of user_segments for which the rule applies, empty applies to all
+pub struct EqCondData {
+    pub name: String,
+    pub value: Vec<String>,
     #[serde(default)]
-    pub user_segments: Vec<LowerCaseString>,
-    #[serde(default)]
-    pub environments: Vec<LowerCaseString>,
-    /// The sampling rate for trace matching this rule
-    pub sample_rate: f64,
-    /// Specifies to what type of item does this rule apply
-    pub strategy: SamplingStrategy,
+    pub ignore_case: bool,
 }
 
-#[derive(Debug, Clone, Serialize, Default)]
-pub struct LowerCaseString(String);
-
-impl LowerCaseString {
-    pub fn new(val: &str) -> LowerCaseString {
-        LowerCaseString(val.to_lowercase())
-    }
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct GlobCondData {
+    pub name: String,
+    pub value: GlobPatterns,
 }
 
-impl From<&str> for LowerCaseString {
-    fn from(val: &str) -> Self {
-        LowerCaseString(val.to_lowercase())
-    }
+/// Keeps inner conditions for combinator conditions
+/// This structure is used to aid the serialisation of Rules.
+/// Since we use internally tagged serialisation for RuleConditions
+/// We need to add an inner level when serializing in order not to have a
+/// clash fo the tag operator between the inner and the outer Conditions
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct InnerConditions {
+    inner: Vec<RuleCondition>,
 }
 
-impl PartialEq for LowerCaseString {
-    fn eq(&self, other: &Self) -> bool {
-        self.0 == other.0
-    }
+/// Keeps inner conditions for combinator conditions with one element (i.e. Not)
+/// This structure is used to aid the serialisation of Rules.
+/// See [InnerConditions] for further explanations.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct InnerNotCondition {
+    inner: Box<RuleCondition>,
 }
 
-impl Eq for LowerCaseString {}
-
-struct LowerCaseStringVisitor;
-
-impl<'de> Visitor<'de> for LowerCaseStringVisitor {
-    type Value = LowerCaseString;
+/// A condition from a sampling rule
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase", tag = "op")]
+pub enum RuleCondition {
+    Eq(EqCondData),
+    Glob(GlobCondData),
+    Or(InnerConditions),
+    And(InnerConditions),
+    Not(InnerNotCondition),
+    #[serde(other)]
+    Unsupported,
+}
 
-    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
-        formatter.write_str("a string to be converted to a lowercase string ")
+impl RuleCondition {
+    /// Checks if Relay supports this condition (in other words if the condition had any unknown configuration
+    /// which was serialized as "Unsupported" (because the configuration is either faulty or was created for a
+    /// newer relay that supports some other condition types)
+    fn is_not_supported(&self) -> bool {
```

Comment:
The doc string says that it checks if the rule is supported, but the function checks the inverse. Can we simplify this and change the method to be `supported()`? The `is_` prefix can be removed or can stay based on your preference).

Reply:
Ok, It would have been strange to say "Checks if Relay doesn't support this condition", but you are right `supported()` is more straightforward.